app/routes/app/brainstorm/new.tsx

```diff
@@ -21,37 +26,95 @@ export const loader = async () => {
 
 const validator = withZod(LaborMarketNewSchema);
 
-export const action = async ({ request }: DataFunctionArgs) => {
-  const data = await validator.validate(await request.formData());
-  if (data.error) return validationError(data.error);
+type ActionResponse = { preparedLaborMarket: LaborMarketPrepared } | ValidationErrorResponseData;
+export const action = async ({ request }: ActionArgs) => {
+  const result = await validator.validate(await request.formData());
+  if (result.error) return validationError(result.error);
+
+  // Additional server side validation
+  if (result.data.launchAccess === "delegates") {
+    const badgerDataMissing = !result.data.launchBadgerAddress || !result.data.launchBadgerTokenId;
+    if (badgerDataMissing) {
+      return validationError({
+        fieldErrors: {
+          launchBadgerAddress: !result.data.launchBadgerAddress
+            ? "Required if delegates selected for launch access"
+            : "",
+          launchBadgerTokenId: !result.data.launchBadgerTokenId
+            ? "Required if delegates selected for launch access"
+            : "",
+        },
+      });
+    }
+  }
 
-  const prepared = await prepareLaborMarket(data.data);
-  return typedjson({ prepared });
+  const preparedLaborMarket = await prepareLaborMarket(result.data);
+  return typedjson({ preparedLaborMarket });
 };
 
 export default function CreateMarketplace() {
   const transition = useTransition();
   const { projects, tokens } = useTypedLoaderData<typeof loader>();
+  const actionData = useTypedActionData<ActionResponse>();
+
+  const [modalData, setModalData] = useState<{ laborMarket?: LaborMarketPrepared; isOpen: boolean }>({ isOpen: false });
+
+  function closeModal() {
+    setModalData((previousInputs) => ({ ...previousInputs, isOpen: false }));
+  }
+
+  useEffect(() => {
+    if (actionData && "preparedLaborMarket" in actionData) {
+      setModalData({ laborMarket: actionData.preparedLaborMarket, isOpen: true });

```

Comment:
Shouldn't have preparedLaborMarket to its own state since its already in actionData. We could just pass `actionData.preparedLaborMarket` to the modal directly. Rather have to do a null check in the component than add another state.

Reply:
issue there is that actionData could either be the validation error type or the prepared labor market type. So need to find a way to constrain it before sending off to the modal.

Reply:
Hmm, could we narrow the type in the render?

```ts
<Modal title="Create Marketplace?" isOpen={modalData.isOpen} onClose={closeModal}>
  { actionData.preparedLaborMarket ? (
    <ConfirmTransaction laborMarket={modalData.laborMarket} onCancel={closeModal} />
  ) : null }
</Modal>
```